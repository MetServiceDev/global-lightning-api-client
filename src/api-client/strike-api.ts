import { Interval } from 'luxon';
import 'isomorphic-fetch';
import {
	StrikeCollection,
	StrikeCollectionType,
	KMLStrikeCollection,
	KML,
	CSVStrikeCollection,
	StrikeCollections,
	GeoJsonStrikeCollection,
	BlitzenStrikeCollection,
	CSV,
} from './strike-collections';
import parseLinkHeader from 'parse-link-header';
import { LightningFeatureCollectionV3, LightningFeatureCollectionV2, BlitzenCollectionV3, BlitzenCollectionV2, BlitzenCollectionV1 } from './model/models';

const MAXIMUM_NUMBER_OF_ATTEMPTS = 3;

export enum LightningDataNetworkProvider {
	toa = 'toa',
	transpower = 'transpower',
	mock = 'mock',
}
export enum LightningStrikeDirection {
	CLOUD = 'CLOUD',
	GROUND = 'GROUND',
}

/**
 * NOTE: This library does not support vector tiles as those are lossy and better suited to visualisation than data aggregation.
 * We also do not support GeoBufs at this point in time
 */
export enum SupportedMimeType {
	// GeoBuf = 'application/octet-stream',
	KML = 'application/vnd.google-earth.kml+xml',
	CSV = 'text/csv',
	Blitzen = 'application/vnd.metraweather.blitzen',
	BlitzenV3 = 'application/vnd.metraweather.blitzen.v3',
	BlitzenV2 = 'application/vnd.metraweather.blitzen.v2',
	BlitzenV1 = 'application/vnd.metraweather.blitzen.v1',
	GeoJson = 'application/vnd.geo+json',
	GeoJsonV3 = 'application/vnd.metraweather.lightning.geo+json.v3',
	GeoJsonV2 = 'application/vnd.metraweather.lightning.geo+json.v2',
}

export enum SupportedVersion {
	Four = 'v4',
}
export enum CredentialType {
	jwt = 'jwt',
	apiKey = 'apiKey',
}

const assertUnreachable = (x: never, customErrorMessage: string = `Unreachable code reached, handling for value '${x}' is not defined.`): never => {
	throw new Error(customErrorMessage);
};

const getAuthorizationPrefix = (type: CredentialType) => {
	switch (type) {
		case CredentialType.jwt: {
			return 'Bearer';
		}
		case CredentialType.apiKey: {
			return 'ApiKey';
		}
		default: {
			return assertUnreachable(type, `Credential Type '${type}' is not understood/supported.`);
		}
	}
};

export type Bbox = [number, number, number, number];

export interface StrikeQueryParameters {
	credentials: {
		type: CredentialType;
		token: string;
	};
	apiVersion: SupportedVersion;
	time: Interval;
	bbox: Bbox;
	limit: number;
	offset: number;
	providers?: LightningDataNetworkProvider[];
	directions?: LightningStrikeDirection[];
}
export interface ApiResponse<SC extends StrikeCollectionType> {
	strikeCollection: StrikeCollection<SC>;
	strikesRemaining: boolean;
}

async function fetchAndFormatStrikes(format: SupportedMimeType.GeoJson, params: StrikeQueryParameters): Promise<ApiResponse<LightningFeatureCollectionV3>>;
async function fetchAndFormatStrikes(format: SupportedMimeType.GeoJsonV3, params: StrikeQueryParameters): Promise<ApiResponse<LightningFeatureCollectionV3>>;
async function fetchAndFormatStrikes(format: SupportedMimeType.GeoJsonV2, params: StrikeQueryParameters): Promise<ApiResponse<LightningFeatureCollectionV2>>;
async function fetchAndFormatStrikes(format: SupportedMimeType.Blitzen, params: StrikeQueryParameters): Promise<ApiResponse<BlitzenCollectionV3>>;
async function fetchAndFormatStrikes(format: SupportedMimeType.BlitzenV3, params: StrikeQueryParameters): Promise<ApiResponse<BlitzenCollectionV3>>;
async function fetchAndFormatStrikes(format: SupportedMimeType.BlitzenV2, params: StrikeQueryParameters): Promise<ApiResponse<BlitzenCollectionV2>>;
async function fetchAndFormatStrikes(format: SupportedMimeType.BlitzenV1, params: StrikeQueryParameters): Promise<ApiResponse<BlitzenCollectionV1>>;
async function fetchAndFormatStrikes(format: SupportedMimeType.CSV, params: StrikeQueryParameters): Promise<ApiResponse<CSV>>;
async function fetchAndFormatStrikes(format: SupportedMimeType.KML, params: StrikeQueryParameters): Promise<ApiResponse<KML>>;
async function fetchAndFormatStrikes(format: SupportedMimeType, params: StrikeQueryParameters): Promise<ApiResponse<StrikeCollectionType>>;
async function fetchAndFormatStrikes(
	format: SupportedMimeType,
	{ credentials, apiVersion, time, bbox, limit, offset, providers, directions }: StrikeQueryParameters
): Promise<ApiResponse<StrikeCollectionType>> {
	const providerString = providers ? `&provider=${providers.join(',')}` : '';
	const directionsString = directions ? `&direction=${directions.join(',')}` : '';
	const timeString = `${time.start.setZone('utc').toISO()}--${time.end.setZone('utc').toISO()}`;
	const url = `https://lightning.api.metraweather.com/${apiVersion}/strikes?time=${timeString}&bbox=${bbox}&limit=${limit}&offset=${offset}${providerString}${directionsString}`;
	// console.log(`Making request for ${url} with Accept: "${format}"`);
	const response = await fetch(url, {
		headers: {
			Accept: `${format}`,
			Authorization: `${getAuthorizationPrefix(credentials.type)} ${credentials.token}`,
		},
	});
	let strikesRemaining = false;
	const linkHeader = response.headers.get('link');
	if (linkHeader !== null) {
		const links = parseLinkHeader(linkHeader);
		strikesRemaining = links?.next !== undefined && links?.next !== null;
	}
	// console.log(`Strikes remaining: ${strikesRemaining}`);
	const strikeCollection = await getFormattedStrikes(format, response);
	return {
		strikeCollection,
		strikesRemaining,
	};
}
async function fetchAndFormatStrikesAndFormatRetryingOnFail(
	format: SupportedMimeType.GeoJson,
	params: StrikeQueryParameters
): Promise<ApiResponse<LightningFeatureCollectionV3>>;
async function fetchAndFormatStrikesAndFormatRetryingOnFail(
	format: SupportedMimeType.GeoJsonV3,
	params: StrikeQueryParameters
): Promise<ApiResponse<LightningFeatureCollectionV3>>;
async function fetchAndFormatStrikesAndFormatRetryingOnFail(
	format: SupportedMimeType.GeoJsonV2,
	params: StrikeQueryParameters
): Promise<ApiResponse<LightningFeatureCollectionV2>>;
async function fetchAndFormatStrikesAndFormatRetryingOnFail(
	format: SupportedMimeType.Blitzen,
	params: StrikeQueryParameters
): Promise<ApiResponse<BlitzenCollectionV3>>;
async function fetchAndFormatStrikesAndFormatRetryingOnFail(
	format: SupportedMimeType.BlitzenV3,
	params: StrikeQueryParameters
): Promise<ApiResponse<BlitzenCollectionV3>>;
async function fetchAndFormatStrikesAndFormatRetryingOnFail(
	format: SupportedMimeType.BlitzenV2,
	params: StrikeQueryParameters
): Promise<ApiResponse<BlitzenCollectionV2>>;
async function fetchAndFormatStrikesAndFormatRetryingOnFail(
	format: SupportedMimeType.BlitzenV1,
	params: StrikeQueryParameters
): Promise<ApiResponse<BlitzenCollectionV1>>;
async function fetchAndFormatStrikesAndFormatRetryingOnFail(format: SupportedMimeType.CSV, params: StrikeQueryParameters): Promise<ApiResponse<CSV>>;
async function fetchAndFormatStrikesAndFormatRetryingOnFail(format: SupportedMimeType.KML, queryParameters: StrikeQueryParameters): Promise<ApiResponse<KML>>;
async function fetchAndFormatStrikesAndFormatRetryingOnFail<T extends StrikeCollectionType>(
	format: SupportedMimeType,
	queryParameters: StrikeQueryParameters
): Promise<ApiResponse<T>>;
async function fetchAndFormatStrikesAndFormatRetryingOnFail(
	format: SupportedMimeType,
	queryParameters: StrikeQueryParameters
): Promise<ApiResponse<StrikeCollectionType>> {
	let numberOfAttempts = 0;
	while (true) {
		try {
			return await fetchAndFormatStrikes(format, queryParameters);
		} catch (error) {
			numberOfAttempts++;
			await sleepForBackoffTime(numberOfAttempts);
			if (numberOfAttempts === MAXIMUM_NUMBER_OF_ATTEMPTS) {
				throw error;
			}
		}
	}
}
/**
 * Waits for an additional 5 seconds for each number of attempts that have failed
 */
const sleepForBackoffTime = async (numberOfAttempts: number) => {
	const numberOfSecondsToWait = 5 * numberOfAttempts;
	new Promise((resolve) => setTimeout(resolve, numberOfSecondsToWait * 1000));
};

async function getFormattedStrikes(format: SupportedMimeType.GeoJson, response: Response): Promise<GeoJsonStrikeCollection<LightningFeatureCollectionV3>>;
async function getFormattedStrikes(format: SupportedMimeType.GeoJsonV3, response: Response): Promise<GeoJsonStrikeCollection<LightningFeatureCollectionV3>>;
async function getFormattedStrikes(format: SupportedMimeType.GeoJsonV2, response: Response): Promise<GeoJsonStrikeCollection<LightningFeatureCollectionV2>>;
async function getFormattedStrikes(format: SupportedMimeType.Blitzen, response: Response): Promise<BlitzenStrikeCollection<BlitzenCollectionV3>>;
async function getFormattedStrikes(format: SupportedMimeType.BlitzenV3, response: Response): Promise<BlitzenStrikeCollection<BlitzenCollectionV3>>;
async function getFormattedStrikes(format: SupportedMimeType.BlitzenV2, response: Response): Promise<BlitzenStrikeCollection<BlitzenCollectionV2>>;
async function getFormattedStrikes(format: SupportedMimeType.BlitzenV1, response: Response): Promise<BlitzenStrikeCollection<BlitzenCollectionV1>>;
async function getFormattedStrikes(format: SupportedMimeType.KML, response: Response): Promise<KMLStrikeCollection>;
async function getFormattedStrikes(format: SupportedMimeType.CSV, response: Response): Promise<CSVStrikeCollection>;
async function getFormattedStrikes<T extends StrikeCollectionType>(format: SupportedMimeType, response: Response): Promise<StrikeCollections>;
async function getFormattedStrikes(format: SupportedMimeType, response: Response): Promise<StrikeCollections> {
	switch (format) {
		case SupportedMimeType.GeoJson: {
			// Fallthrough
		}
		case SupportedMimeType.GeoJsonV3: {
			return new GeoJsonStrikeCollection<LightningFeatureCollectionV3>(response);
		}
		case SupportedMimeType.GeoJsonV2: {
			return new GeoJsonStrikeCollection<LightningFeatureCollectionV2>(response);
		}
		case SupportedMimeType.Blitzen: {
			// Fallthrough
		}
		case SupportedMimeType.BlitzenV3: {
			return new BlitzenStrikeCollection<BlitzenCollectionV3>(response);
		}
		case SupportedMimeType.BlitzenV2: {
			return new BlitzenStrikeCollection<BlitzenCollectionV2>(response);
		}
		case SupportedMimeType.BlitzenV1: {
			return new BlitzenStrikeCollection<BlitzenCollectionV1>(response);
		}
		case SupportedMimeType.KML: {
			return new KMLStrikeCollection(response);
		}
		case SupportedMimeType.CSV: {
			return new CSVStrikeCollection(response);
		}
		default: {
			return assertUnreachable(format, `Format '${format}' is not supported.`);
		}
		// case SupportedMimeType.GeoBuf: {
		// TODO: The API does not appear to implement GeoBufs correctly. They are more than just a buffer of JSON.
		// 	geobuf.decode(new Pbf(data));
		// }
	}
}

export { getAuthorizationPrefix, fetchAndFormatStrikes, getFormattedStrikes, fetchAndFormatStrikesAndFormatRetryingOnFail };
