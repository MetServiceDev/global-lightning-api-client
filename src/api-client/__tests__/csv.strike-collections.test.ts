import { getFormattedStrikes, SupportedMimeType } from '../strike-api';
import {
	CSV_RESPONSE_ONE,
	CSV_RESPONSE_TWO,
	CSV_RESPONSE_THREE,
	CSV_EMPTY_RESPONSE,
	CSV_RESPONSE_TWO_WITH_DIFFERENT_HEADERS,
} from './strikes-collections-test-resources';

describe('When parsing CSV', () => {
	it('should return valid CSV', async () => {
		const strikeCollection = await getFormattedStrikes(SupportedMimeType.CSV, {
			text: () => Promise.resolve(CSV_RESPONSE_ONE),
		} as Response);
		const formattedStrikes = await strikeCollection.collection;
		expect(formattedStrikes.body[0]).toEqual('105.911369,3.79772,2020-06-20T00:00:00.323Z,-13.6,GROUND,toa,1592611200.323,1,-12,0.25,0.25');
		expect(formattedStrikes).toMatchSnapshot();
	});
	it('should be able to reformat to the same string', async () => {
		const strikeCollection = await getFormattedStrikes(SupportedMimeType.CSV, {
			text: () => Promise.resolve(CSV_RESPONSE_ONE),
		} as Response);
		const reformattedString = await strikeCollection.toString();
		expect(reformattedString).toEqual(CSV_RESPONSE_ONE);
	});
	it('should be able to merge two collections', async () => {
		const strikeCollection = await getFormattedStrikes(SupportedMimeType.CSV, {
			text: () => Promise.resolve(CSV_RESPONSE_ONE),
		} as Response);
		const strikeCollectionTwo = await getFormattedStrikes(SupportedMimeType.CSV, {
			text: () => Promise.resolve(CSV_RESPONSE_TWO),
		} as Response);
		const mergedCollection = await strikeCollection.mergeCollection(strikeCollectionTwo);
		// const formattedStrikes = await strikeCollection.collection;
		expect(mergedCollection.body[0]).toEqual('105.911369,3.79772,2020-06-20T00:00:00.323Z,-13.6,GROUND,toa,1592611200.323,1,-12,0.25,0.25');
		expect(mergedCollection).toMatchSnapshot();
	});
	it.skip('should merge collections in order', () => {});
	it.skip('should be able to calculate any new strikes', () => {});
	it('should be able to merge multiple collections', async () => {
		const strikeCollection = await getFormattedStrikes(SupportedMimeType.CSV, {
			text: () => Promise.resolve(CSV_RESPONSE_ONE),
		} as Response);
		const strikeCollectionTwo = await getFormattedStrikes(SupportedMimeType.CSV, {
			text: () => Promise.resolve(CSV_RESPONSE_TWO),
		} as Response);
		const strikeCollectionThree = await getFormattedStrikes(SupportedMimeType.CSV, {
			text: () => Promise.resolve(CSV_RESPONSE_THREE),
		} as Response);
		strikeCollection.mergeCollections(strikeCollectionTwo, strikeCollectionThree);
		const formattedStrikes = await strikeCollection.collection;
		expect(formattedStrikes.body[0]).toEqual('105.911369,3.79772,2020-06-20T00:00:00.323Z,-13.6,GROUND,toa,1592611200.323,1,-12,0.25,0.25');
		expect(formattedStrikes).toMatchSnapshot();
	});
	it('should be able to merge an empty collection', async () => {
		const strikeCollection = await getFormattedStrikes(SupportedMimeType.CSV, {
			text: () => Promise.resolve(CSV_RESPONSE_ONE),
		} as Response);
		const emptyCollection = await getFormattedStrikes(SupportedMimeType.CSV, {
			text: () => Promise.resolve(CSV_EMPTY_RESPONSE),
		} as Response);
		strikeCollection.mergeCollections(emptyCollection);
		expect(await strikeCollection.toString()).toEqual(CSV_RESPONSE_ONE);
		const formattedStrikes = await strikeCollection.collection;
		expect(formattedStrikes.body[0]).toEqual('105.911369,3.79772,2020-06-20T00:00:00.323Z,-13.6,GROUND,toa,1592611200.323,1,-12,0.25,0.25');
	});
	it('should be able to merge two empty collections', async () => {
		const strikeCollection = await getFormattedStrikes(SupportedMimeType.CSV, {
			text: () => Promise.resolve(CSV_EMPTY_RESPONSE),
		} as Response);
		const emptyCollection = await getFormattedStrikes(SupportedMimeType.CSV, {
			text: () => Promise.resolve(CSV_EMPTY_RESPONSE),
		} as Response);
		strikeCollection.mergeCollections(emptyCollection);
		const formattedStrikes = await strikeCollection.collection;
		expect(formattedStrikes.body).toEqual([]);
		expect(await strikeCollection.toString()).toEqual(CSV_EMPTY_RESPONSE);
	});
	it('should be able to handle out of order CSV headers when merging', async () => {
		const strikeCollection = await getFormattedStrikes(SupportedMimeType.CSV, {
			text: () => Promise.resolve(CSV_RESPONSE_ONE),
		} as Response);
		const strikeCollectionTwo = await getFormattedStrikes(SupportedMimeType.CSV, {
			text: () => Promise.resolve(CSV_RESPONSE_TWO_WITH_DIFFERENT_HEADERS),
		} as Response);
		const mergedCollection = await strikeCollection.mergeCollection(strikeCollectionTwo);
		// const formattedStrikes = await strikeCollection.collection;
		expect(mergedCollection.body[0]).toEqual('105.911369,3.79772,2020-06-20T00:00:00.323Z,-13.6,GROUND,toa,1592611200.323,1,-12,0.25,0.25');
		expect(mergedCollection.body[10]).not.toEqual('0.25,-76.516297,39.093173,-6.4,GROUND,toa,1592611200.629,1,9,0.25,2020-06-20T00:00:00.629Z');
		expect(mergedCollection.body[10]).toEqual('-80.874821,34.163358,2020-06-20T00:01:00.385Z,0,CLOUD,toa,1592611260.385,1,-74,0.5,0.25');
		expect(mergedCollection).toMatchSnapshot();
	});
});
