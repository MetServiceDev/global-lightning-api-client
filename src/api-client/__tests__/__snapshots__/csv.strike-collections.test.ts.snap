// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When parsing CSV should be able to handle out of order CSV headers when merging 1`] = `
Object {
  "body": Array [
    "105.911369,3.79772,2020-06-20T00:00:00.323Z,-13.6,GROUND,toa,1592611200.323,1,-12,0.25,0.25",
    "-76.645326,39.344666,2020-06-20T00:00:00.628Z,0,CLOUD,toa,1592611200.628,1,-49,0.25,0.25",
    "-76.516297,39.093173,2020-06-20T00:00:00.629Z,-6.4,GROUND,toa,1592611200.629,1,9,0.25,0.25",
    "-76.751111,39.359796,2020-06-20T00:00:00.631Z,0,CLOUD,toa,1592611200.631,1,-28,0.5,0.25",
    "-76.699283,39.357346,2020-06-20T00:00:00.631Z,0,CLOUD,toa,1592611200.631,6,-80,1.75,0.25",
    "-76.627881,39.360246,2020-06-20T00:00:00.631Z,-66.2,GROUND,toa,1592611200.631,1,-41,0.25,0.25",
    "-76.623955,39.355165,2020-06-20T00:00:00.631Z,-65,GROUND,toa,1592611200.631,1,82,0.5,0.25",
    "-76.666255,39.283483,2020-06-20T00:00:00.653Z,0,CLOUD,toa,1592611200.653,1,-52,0.25,0.25",
    "-76.568837,39.355049,2020-06-20T00:00:00.653Z,-20.3,GROUND,toa,1592611200.653,1,-1,0.25,0.25",
    "-76.522296,39.390466,2020-06-20T00:00:00.653Z,0,CLOUD,toa,1592611200.653,1,-74,0.5,0.25",
    "-80.874821,34.163358,2020-06-20T00:01:00.385Z,0,CLOUD,toa,1592611260.385,1,-74,0.5,0.25",
    "-82.357371,30.250675,2020-06-20T00:01:00.552Z,-26,GROUND,toa,1592611260.552,2,-89,0.75,0.25",
    "-82.353037,30.263559,2020-06-20T00:01:00.552Z,0,CLOUD,toa,1592611260.552,1,83,0.25,0.25",
    "-91.441549,16.24059,2020-06-20T00:01:00.800Z,-41.3,GROUND,toa,1592611260.8,6,11,1.75,0.25",
    "26.773836,59.46069,2020-06-20T00:01:01.000Z,-25.2,GROUND,toa,1592611261,1,7,0.5,0.25",
    "26.923863,59.507565,2020-06-20T00:01:01.028Z,-15.1,GROUND,toa,1592611261.028,3,11,0.75,0.25",
    "26.904206,59.500731,2020-06-20T00:01:01.077Z,-14,GROUND,toa,1592611261.077,3,11,1,0.25",
    "11.790852,56.359199,2020-06-20T00:01:01.296Z,95.3,GROUND,toa,1592611261.296,1,19,0.25,0.25",
    "-101.650011,32.374309,2020-06-20T00:01:01.382Z,0,CLOUD,toa,1592611261.382,16,34,4.25,0.5",
    "-101.366501,32.777182,2020-06-20T00:01:01.382Z,-29.9,GROUND,toa,1592611261.382,1,-24,0.25,0.25",
  ],
  "header": "lon,lat,dateTime,kA,strike_type,source,unixTime,GDOP,ellipse_bearing,ellipse_major_axis,ellipse_minor_axis",
}
`;

exports[`When parsing CSV should be able to merge multiple collections 1`] = `
Object {
  "body": Array [
    "105.911369,3.79772,2020-06-20T00:00:00.323Z,-13.6,GROUND,toa,1592611200.323,1,-12,0.25,0.25",
    "-76.645326,39.344666,2020-06-20T00:00:00.628Z,0,CLOUD,toa,1592611200.628,1,-49,0.25,0.25",
    "-76.516297,39.093173,2020-06-20T00:00:00.629Z,-6.4,GROUND,toa,1592611200.629,1,9,0.25,0.25",
    "-76.751111,39.359796,2020-06-20T00:00:00.631Z,0,CLOUD,toa,1592611200.631,1,-28,0.5,0.25",
    "-76.699283,39.357346,2020-06-20T00:00:00.631Z,0,CLOUD,toa,1592611200.631,6,-80,1.75,0.25",
    "-76.627881,39.360246,2020-06-20T00:00:00.631Z,-66.2,GROUND,toa,1592611200.631,1,-41,0.25,0.25",
    "-76.623955,39.355165,2020-06-20T00:00:00.631Z,-65,GROUND,toa,1592611200.631,1,82,0.5,0.25",
    "-76.666255,39.283483,2020-06-20T00:00:00.653Z,0,CLOUD,toa,1592611200.653,1,-52,0.25,0.25",
    "-76.568837,39.355049,2020-06-20T00:00:00.653Z,-20.3,GROUND,toa,1592611200.653,1,-1,0.25,0.25",
    "-76.522296,39.390466,2020-06-20T00:00:00.653Z,0,CLOUD,toa,1592611200.653,1,-74,0.5,0.25",
    "-80.874821,34.163358,2020-06-20T00:01:00.385Z,0,CLOUD,toa,1592611260.385,1,-74,0.5,0.25",
    "-82.357371,30.250675,2020-06-20T00:01:00.552Z,-26,GROUND,toa,1592611260.552,2,-89,0.75,0.25",
    "-82.353037,30.263559,2020-06-20T00:01:00.552Z,0,CLOUD,toa,1592611260.552,1,83,0.25,0.25",
    "-91.441549,16.24059,2020-06-20T00:01:00.800Z,-41.3,GROUND,toa,1592611260.8,6,11,1.75,0.25",
    "26.773836,59.46069,2020-06-20T00:01:01.000Z,-25.2,GROUND,toa,1592611261,1,7,0.5,0.25",
    "26.923863,59.507565,2020-06-20T00:01:01.028Z,-15.1,GROUND,toa,1592611261.028,3,11,0.75,0.25",
    "26.904206,59.500731,2020-06-20T00:01:01.077Z,-14,GROUND,toa,1592611261.077,3,11,1,0.25",
    "11.790852,56.359199,2020-06-20T00:01:01.296Z,95.3,GROUND,toa,1592611261.296,1,19,0.25,0.25",
    "-101.650011,32.374309,2020-06-20T00:01:01.382Z,0,CLOUD,toa,1592611261.382,16,34,4.25,0.5",
    "-101.366501,32.777182,2020-06-20T00:01:01.382Z,-29.9,GROUND,toa,1592611261.382,1,-24,0.25,0.25",
    "-79.357062,49.79451,2020-06-20T00:02:00.077Z,-56.6,GROUND,toa,1592611320.077,2,23,0.5,0.25",
    "-101.955629,39.873289,2020-06-20T00:02:00.235Z,28.7,GROUND,toa,1592611320.235,1,29,0.25,0.25",
    "-101.995631,39.87736,2020-06-20T00:02:00.289Z,19,GROUND,toa,1592611320.289,1,-11,0.25,0.25",
    "105.181056,3.354772,2020-06-20T00:02:00.981Z,-13.1,GROUND,toa,1592611320.981,1,-9,0.25,0.25",
    "-79.969339,35.098912,2020-06-20T00:02:01.364Z,-18.2,GROUND,toa,1592611321.364,1,-76,0.5,0.25",
    "-80.069913,34.980473,2020-06-20T00:02:01.370Z,-35.9,GROUND,toa,1592611321.37,1,-21,0.25,0.25",
    "105.974937,3.83109,2020-06-20T00:02:01.517Z,-29.8,GROUND,toa,1592611321.517,1,-24,0.25,0.25",
    "105.677619,3.858295,2020-06-20T00:02:01.544Z,0,CLOUD,toa,1592611321.544,1,16,0.5,0.25",
    "105.9039,3.811184,2020-06-20T00:02:01.544Z,-11,GROUND,toa,1592611321.544,1,-8,0.25,0.25",
    "105.678063,3.858444,2020-06-20T00:02:01.561Z,-7.8,GROUND,toa,1592611321.561,1,16,0.5,0.25",
  ],
  "header": "lon,lat,dateTime,kA,strike_type,source,unixTime,GDOP,ellipse_bearing,ellipse_major_axis,ellipse_minor_axis",
}
`;

exports[`When parsing CSV should be able to merge two collections 1`] = `
Object {
  "body": Array [
    "105.911369,3.79772,2020-06-20T00:00:00.323Z,-13.6,GROUND,toa,1592611200.323,1,-12,0.25,0.25",
    "-76.645326,39.344666,2020-06-20T00:00:00.628Z,0,CLOUD,toa,1592611200.628,1,-49,0.25,0.25",
    "-76.516297,39.093173,2020-06-20T00:00:00.629Z,-6.4,GROUND,toa,1592611200.629,1,9,0.25,0.25",
    "-76.751111,39.359796,2020-06-20T00:00:00.631Z,0,CLOUD,toa,1592611200.631,1,-28,0.5,0.25",
    "-76.699283,39.357346,2020-06-20T00:00:00.631Z,0,CLOUD,toa,1592611200.631,6,-80,1.75,0.25",
    "-76.627881,39.360246,2020-06-20T00:00:00.631Z,-66.2,GROUND,toa,1592611200.631,1,-41,0.25,0.25",
    "-76.623955,39.355165,2020-06-20T00:00:00.631Z,-65,GROUND,toa,1592611200.631,1,82,0.5,0.25",
    "-76.666255,39.283483,2020-06-20T00:00:00.653Z,0,CLOUD,toa,1592611200.653,1,-52,0.25,0.25",
    "-76.568837,39.355049,2020-06-20T00:00:00.653Z,-20.3,GROUND,toa,1592611200.653,1,-1,0.25,0.25",
    "-76.522296,39.390466,2020-06-20T00:00:00.653Z,0,CLOUD,toa,1592611200.653,1,-74,0.5,0.25",
    "-80.874821,34.163358,2020-06-20T00:01:00.385Z,0,CLOUD,toa,1592611260.385,1,-74,0.5,0.25",
    "-82.357371,30.250675,2020-06-20T00:01:00.552Z,-26,GROUND,toa,1592611260.552,2,-89,0.75,0.25",
    "-82.353037,30.263559,2020-06-20T00:01:00.552Z,0,CLOUD,toa,1592611260.552,1,83,0.25,0.25",
    "-91.441549,16.24059,2020-06-20T00:01:00.800Z,-41.3,GROUND,toa,1592611260.8,6,11,1.75,0.25",
    "26.773836,59.46069,2020-06-20T00:01:01.000Z,-25.2,GROUND,toa,1592611261,1,7,0.5,0.25",
    "26.923863,59.507565,2020-06-20T00:01:01.028Z,-15.1,GROUND,toa,1592611261.028,3,11,0.75,0.25",
    "26.904206,59.500731,2020-06-20T00:01:01.077Z,-14,GROUND,toa,1592611261.077,3,11,1,0.25",
    "11.790852,56.359199,2020-06-20T00:01:01.296Z,95.3,GROUND,toa,1592611261.296,1,19,0.25,0.25",
    "-101.650011,32.374309,2020-06-20T00:01:01.382Z,0,CLOUD,toa,1592611261.382,16,34,4.25,0.5",
    "-101.366501,32.777182,2020-06-20T00:01:01.382Z,-29.9,GROUND,toa,1592611261.382,1,-24,0.25,0.25",
  ],
  "header": "lon,lat,dateTime,kA,strike_type,source,unixTime,GDOP,ellipse_bearing,ellipse_major_axis,ellipse_minor_axis",
}
`;

exports[`When parsing CSV should return valid CSV 1`] = `
Object {
  "body": Array [
    "105.911369,3.79772,2020-06-20T00:00:00.323Z,-13.6,GROUND,toa,1592611200.323,1,-12,0.25,0.25",
    "-76.645326,39.344666,2020-06-20T00:00:00.628Z,0,CLOUD,toa,1592611200.628,1,-49,0.25,0.25",
    "-76.516297,39.093173,2020-06-20T00:00:00.629Z,-6.4,GROUND,toa,1592611200.629,1,9,0.25,0.25",
    "-76.751111,39.359796,2020-06-20T00:00:00.631Z,0,CLOUD,toa,1592611200.631,1,-28,0.5,0.25",
    "-76.699283,39.357346,2020-06-20T00:00:00.631Z,0,CLOUD,toa,1592611200.631,6,-80,1.75,0.25",
    "-76.627881,39.360246,2020-06-20T00:00:00.631Z,-66.2,GROUND,toa,1592611200.631,1,-41,0.25,0.25",
    "-76.623955,39.355165,2020-06-20T00:00:00.631Z,-65,GROUND,toa,1592611200.631,1,82,0.5,0.25",
    "-76.666255,39.283483,2020-06-20T00:00:00.653Z,0,CLOUD,toa,1592611200.653,1,-52,0.25,0.25",
    "-76.568837,39.355049,2020-06-20T00:00:00.653Z,-20.3,GROUND,toa,1592611200.653,1,-1,0.25,0.25",
    "-76.522296,39.390466,2020-06-20T00:00:00.653Z,0,CLOUD,toa,1592611200.653,1,-74,0.5,0.25",
  ],
  "header": "lon,lat,dateTime,kA,strike_type,source,unixTime,GDOP,ellipse_bearing,ellipse_major_axis,ellipse_minor_axis",
}
`;
